### PR TITLE
follow-up: add check for "ComponentBitCount" to basicxmlmetadata-checker

### PR DESCRIPTION
**Pull request by ptahmose**
## Description

Adding a check for the presence and validity of the field "ComponentBitCount" under Dimensions/Channels/Channel[].
Note: From a file-format perspective, the ComponentBitCount information is optional and can be omitted. However, as it turned out, existing software seems to rely on the presence of this information in some cases. Therefore, for the time being, we choose to report a warning if the information is missing.
We also add a basal check for the validity of the value, based on the stated PixelType. And - we only report a warning if the stated PixelType is an integer-pixel-type.

This PR is following up on #17 .

### Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## How Has This Been Tested?

A testcase has been added, and existing test-results have been adapted.

## Checklist:

- [x] I followed the Contributing Guidelines.
- [x] I did a self-review.
- [x] I commented my code, particularly in hard-to-understand areas.
- [x] I updated the documentation.
- [x] I have added tests that prove my fix is effective or that my feature works.
- [x] New and existing unit tests pass locally with my changes.

### CODE REVIEW

File: CZICheck/checkers/checkerXmlBasicMetadataValidation.cpp [hunk 1]
```diff
@@ -204,8 +207,86 @@ void CCheckBasicMetadataValidation::CheckPixelTypeInformation(const std::shared_
...
             this->result_gatherer_.ReportFinding(finding);
         }
+
+        // check for presence and basal validity of "ComponentBitCount" information
+        int channel_info_component_bit_count;
+        if (!channel_info->TryGetComponentBitCount(&channel_info_component_bit_count))
+        {
+            // Report a warning if no ComponentBitCount information is found in metadata if
+            // * the PixelType information is invalid (i.e. not present)
+            // * or, if the PixelType is an integer pixel type, for which we expect a ComponentBitCount information (or: where it is recommended to have it)
```
Comment (DaveyJonesBitPail) [inline CZICheck/checkers/checkerXmlBasicMetadataValidation.cpp:222]: Is there ever a valid situation where we have no PixelType information but the actual pixel data is not an integer pixel type? Just curious, I learn toward a false positive is better than a false negative especially if it is a warning

File: CZICheck/checkers/checkerXmlBasicMetadataValidation.cpp [hunk 2]
```diff
@@ -204,8 +207,86 @@ void CCheckBasicMetadataValidation::CheckPixelTypeInformation(const std::shared_
...
             CResultGatherer::Finding finding(CCheckBasicMetadataValidation::kCheckType);
             finding.severity = CResultGatherer::Severity::Info;
-            finding.information = "No valid channel-information found in metadata";
+            stringstream ss;
+            ss << "No valid channel pixel_type information found in metadata for channel #" << channelIndex << ".";
+            finding.information = ss.str();
             this->result_gatherer_.ReportFinding(finding);
         }
+
+        // check for presence and basal validity of "ComponentBitCount" information
```
[the remark this answers is not in the document]
Comment (ptahmose) [reply CZICheck/checkers/checkerXmlBasicMetadataValidation.cpp:216]: agreed - and fixed

File: CZICheck/checkers/checkerXmlBasicMetadataValidation.cpp [hunk 3]
```diff
@@ -204,8 +207,86 @@ void CCheckBasicMetadataValidation::CheckPixelTypeInformation(const std::shared_
...
+{
+    switch (pixel_type)  // NOLINT(clang-diagnostic-switch-enum)
+    {
+    case PixelType::Gray8:
+    case PixelType::Bgr24:
+    case PixelType::Bgra32:
+        return component_bit_count > 0 && component_bit_count <= 8;
+    case PixelType::Gray16:
+    case PixelType::Bgr48:
+        return component_bit_count > 0 && component_bit_count <= 16;
```
Comment (DaveyJonesBitPail) [inline CZICheck/checkers/checkerXmlBasicMetadataValidation.cpp:288]: Silly question:
Is the component_bit_count always <= Total Bits Across all Channels / Channel Count? In reasonable pixel types this should be true and we don't support custom esoteric types (afaik)

**Review by DaveyJonesBitPail — APPROVED** [review]

**Comment by DaveyJonesBitPail** [comment]
As evidenced by having to change multiple txt files for expected output, we may want to consider (in the future) to have succinct definitions for expected output for certain czi files so that we only check for certain specific string sequences in output.

**Pull request merged**